File: src/services/ingestion/pipeline.ts
import { Neo4jClient } from "../../db/neo4j-client.js";
import { ContentType, DocumentStatus } from "../../types/enums.js";
import type { Document, Metadata } from "../../types/models.js";
import { EmbeddingService } from "../embedding.js";
import {
  documentTemporalPolicy,
  memoryContentHash,
  withTemporalClass
} from "../memory-policy.js";
import { RelationClassifierService } from "../relation-classifier.js";
import { ChunkingService, type ChunkPayload } from "./chunker.js";
import { MemoryExtractorService, type ExtractedMemory } from "./memory-extractor.js";
import { ImageExtractor } from "./extractors/image.js";
import { PdfExtractor } from "./extractors/pdf.js";
import {
  ConversationExtractor,
  TextExtractor,
  UrlExtractor,
  type Extractor
} from "./extractors/index.js";

export type PipelineInput = {
  title: string;
  contentType: ContentType;
  rawContent: string;
  containerTag: string;
  metadata?: Metadata;
  sourceUrl?: string;
  filePath?: string;
};

export type PipelineResult = {
  document: Document;
  chunkCount: number;
  /** Memories actually created. */
  memoryCount: number;
  /** Extracted memories skipped because an active memory with the same normalised content exists. */
  duplicateCount: number;
  /** Extracted memories skipped because the document's temporal_class requires a validTo none was found. */
  rejectedCount: number;
};

/** Document metadata keys written by the pipeline with the outcome of the memory policy. */
export const PIPELINE_RESULT_METADATA_KEYS = {
  created: "memories_created",
  duplicate: "memories_duplicate",
  rejected: "memories_rejected"
} as const;

/** End-to-end multimodal ingestion orchestrator. */
export class IngestionPipeline {
  private readonly neo4jClient: Neo4jClient;
  private readonly embeddingService: EmbeddingService;
  private readonly relationClassifierService: RelationClassifierService;
  private readonly chunkingService: ChunkingService;
  private readonly memoryExtractorService: MemoryExtractorService;

  /**
   * Creates the ingestion pipeline.
   */
  public constructor(
    neo4jClient: Neo4jClient,
    embeddingService: EmbeddingService,
    relationClassifierService: RelationClassifierService,
    memoryExtractorService: MemoryExtractorService
  ) {
    this.neo4jClient = neo4jClient;
    this.embeddingService = embeddingService;
    this.relationClassifierService = relationClassifierService;
    this.chunkingService = new ChunkingService();
    this.memoryExtractorService = memoryExtractorService;
  }

  /**
   * Creates a document and runs full ingestion.
   * @param input Ingestion input.
   */
  public async ingest(input: PipelineInput): Promise<PipelineResult> {
    const document = await this.neo4jClient.createDocument(input);
    return this.processDocument(document.id);
  }

  /**
   * Runs pipeline stages for an existing document.
   * @param documentId Document id.
   */
  public async processDocument(documentId: string): Promise<PipelineResult> {
    const document = await this.neo4jClient.getDocument(documentId);
    if (!document) {
      throw new Error(`Document not found: ${documentId}`);
    }

    try {
      await this.setStatus(document.id, DocumentStatus.Extracting);
      const extractor = this.getExtractor(document.contentType);
      const extractedText = await extractor.extract(document);
      const extractedDoc = await this.neo4jClient.updateDocument(document.id, {
        rawContent: extractedText,
        status: DocumentStatus.Extracting
      });

      await this.setStatus(document.id, DocumentStatus.Chunking);
      const chunks = this.chunkingService.chunk(
        {
          ...document,
          rawContent: extractedText
        },
        extractedText
      );

      await this.setStatus(document.id, DocumentStatus.ExtractingMemories);
      const filterPrompt = await this.neo4jClient.getContainerFilterPrompt(document.containerTag);
      const extractedMemories = await this.extractMemories(chunks, filterPrompt);
      // Same rules as create_memory: validTo mandatory for temporal intelligence and no exact duplicates,
      // applied before spending embeddings on memories that will not be stored.
      const { accepted, duplicateCount, rejectedCount } = await this.applyMemoryPolicy(document, extractedMemories);

      await this.setStatus(document.id, DocumentStatus.Embedding);
      const chunkEmbeddings = await this.embeddingService.generateEmbeddings(
        chunks.map((chunk) => chunk.content)
      );
      const memoryEmbeddings = accepted.length === 0
        ? []
        : await this.embeddingService.generateEmbeddings(accepted.map((memory) => memory.content));

      await this.setStatus(document.id, DocumentStatus.Indexing);
      if (chunks.length > 0) {
        await this.neo4jClient.createChunks(
          chunks.map((chunk, index) => ({
            content: chunk.content,
            chunkIndex: chunk.chunkIndex,
            containerTag: document.containerTag,
            sourceDocId: document.id,
            metadata: typeof chunk.metadata === "object" ? JSON.stringify(chunk.metadata) : (chunk.metadata ?? ""),
            embedding: chunkEmbeddings[index] ?? []
          }))
        );
      }

      let createdCount = 0;
      for (let i = 0; i < accepted.length; i += 1) {
        const acceptedMemory = accepted[i];
        if (!acceptedMemory) {
          continue;
        }

        const embedding = memoryEmbeddings[i];
        if (!embedding) {
          continue;
        }

        const memory = await this.neo4jClient.createMemory({
          content: acceptedMemory.content,
          memoryType: acceptedMemory.memoryType,
          containerTag: document.containerTag,
          confidence: acceptedMemory.confidence,
          metadata: acceptedMemory.metadata,
          validFrom: acceptedMemory.validFrom ?? undefined,
          validTo: acceptedMemory.validTo ?? undefined,
          sourceDocId: document.id,
          embedding
        });
        createdCount += 1;

        try {
          await this.relationClassifierService.classifyAndApply(memory);
        } catch (e) {
          console.warn(`[pipeline] RelationClassifier skipped for memory ${memory.id}:`, (e as Error).message);
        }
      }

      const finalDocument = await this.neo4jClient.updateDocument(document.id, {
        status: DocumentStatus.Done,
        rawContent: extractedText,
        metadata: {
          ...document.metadata,
          [PIPELINE_RESULT_METADATA_KEYS.created]: createdCount,
          [PIPELINE_RESULT_METADATA_KEYS.duplicate]: duplicateCount,
          [PIPELINE_RESULT_METADATA_KEYS.rejected]: rejectedCount
        }
      });

      if (!finalDocument && !extractedDoc) {
        throw new Error(`Failed to update final status for document ${document.id}`);
      }

      return {
        document: finalDocument ?? extractedDoc ?? document,
        chunkCount: chunks.length,
        memoryCount: createdCount,
        duplicateCount,
        rejectedCount
      };
    } catch (error) {
      const message = error instanceof Error ? error.message : "Unknown pipeline error";
      await this.neo4jClient.updateDocument(document.id, {
        status: DocumentStatus.Error,
        metadata: {
          ...document.metadata,
          pipelineError: message,
          pipelineErrorAt: new Date().toISOString()
        }
      });
      throw error;
    }
  }

  /**
   * Applies the memory policy to extracted memories: inherits the document's default validTo,
   * rejects temporal intelligence without validTo and drops exact duplicates (within the document
   * and against active memories already stored in the container).
   * @param document Source document with temporal policy metadata.
   * @param extractedMemories Memories proposed by the extractor.
   */
  private async applyMemoryPolicy(
    document: Document,
    extractedMemories: ExtractedMemory[]
  ): Promise<{
    accepted: Array<ExtractedMemory & { metadata: Metadata }>;
    duplicateCount: number;
    rejectedCount: number;
  }> {
    const policy = documentTemporalPolicy(document.metadata);
    const seenHashes = new Set<string>();
    const accepted: Array<ExtractedMemory & { metadata: Metadata }> = [];
    let duplicateCount = 0;
    let rejectedCount = 0;

    for (const extracted of extractedMemories) {
      const validTo = extracted.validTo ?? policy.validTo;
      if (policy.temporalClass !== "none" && !validTo) {
        rejectedCount += 1;
        console.warn(
          `[pipeline] Rejected memory without validTo for temporal_class=${policy.temporalClass} in document ${document.id}`
        );
        continue;
      }

      const contentHash = memoryContentHash(document.containerTag, extracted.content);
      if (seenHashes.has(contentHash)) {
        duplicateCount += 1;
        continue;
      }
      const existing = await this.neo4jClient.findActiveMemoryByContentHash(document.containerTag, contentHash);
      if (existing) {
        duplicateCount += 1;
        console.info(`[pipeline] Skipped duplicate of memory ${existing.id} in document ${document.id}`);
        continue;
      }

      seenHashes.add(contentHash);
      accepted.push({
        ...extracted,
        validTo,
        metadata: withTemporalClass(undefined, policy.temporalClass)
      });
    }

    return { accepted, duplicateCount, rejectedCount };
  }

  private async extractMemories(
    chunks: ChunkPayload[],
    filterPrompt: string | null
  ): Promise<ExtractedMemory[]> {
    const allMemories: ExtractedMemory[] = [];

    for (const chunk of chunks) {
      const memories = await this.memoryExtractorService.extractFromChunk(chunk.content, {
        filterPrompt
      });
      allMemories.push(...memories);
    }

    return allMemories;
  }

  private getExtractor(contentType: ContentType): Extractor {
    if (contentType === ContentType.Url) {
      return new UrlExtractor();
    }

    if (contentType === ContentType.Pdf) {
      return new PdfExtractor();
    }

    if (contentType === ContentType.Image) {
      return new ImageExtractor();
    }

    if (contentType === ContentType.Conversation) {
      return new ConversationExtractor();
    }

    return new TextExtractor();
  }

  private async setStatus(documentId: string, status: DocumentStatus): Promise<void> {
    await this.neo4jClient.updateDocument(documentId, { status });
  }

  /**
   * Sets the configuration for a container, including filter prompts.
   * @param containerTag The tag of the container.
   * @param config The configuration object.
   */
  public async set_container_config(containerTag: string, config: { filterPrompt?: string | null }): Promise<void> {
    // Assuming Neo4jClient will have a method to set container configuration.
    // If this method does not exist, it will need to be added to Neo4jClient.
    await this.neo4jClient.setContainerConfig(containerTag, config.filterPrompt ?? null);
  }
}

File: src/services/ingestion/pipeline.py
import asyncio
from typing import List, Optional

from src.db.neo4j_client import Neo4jClient
from src.services.embedding import EmbeddingService
from src.services.ingestion.batching import (
    MemoryBatchInput,
    batch_create_memories,
    gather_with_limit,
)
from src.services.ingestion.chunker import ChunkPayload, ChunkingService
from src.services.ingestion.extractors import (
    ConversationExtractor,
    TextExtractor,
    UrlExtractor,
)
from src.services.ingestion.extractors.base import Extractor
from src.services.ingestion.memory_extractor import (
    ExtractedMemory,
    MemoryExtractorService,
)
from src.services.relation_classifier import RelationClassifierService
from src.types.enums import ContentType, DocumentStatus
from src.types.models import Document, Metadata


class PipelineInput:
    def __init__(
        self,
        title: str,
        content_type: ContentType,
        raw_content: str,
        container_tag: str,
        metadata: Optional[Metadata] = None,
        source_url: Optional[str] = None,
        file_path: Optional[str] = None,
    ):
        self.title = title
        self.content_type = content_type
        self.raw_content = raw_content
        self.container_tag = container_tag
        self.metadata = metadata
        self.source_url = source_url
        self.file_path = file_path


class IngestionPipeline:
    """End-to-end multimodal ingestion orchestrator with parallel LLM processing."""

    def __init__(
        self,
        neo4j_client: Neo4jClient,
        embedding_service: EmbeddingService,
        relation_classifier_service: RelationClassifierService,
        memory_extractor_service: MemoryExtractorService,
    ):
        self.neo4j_client = neo4j_client
        self.embedding_service = embedding_service
        self.relation_classifier_service = relation_classifier_service
        self.chunking_service = ChunkingService()
        self.memory_extractor_service = memory_extractor_service

    async def ingest(
        self, input_data: PipelineInput
    ) -> dict[str, Document | int]:
        """Creates a document and runs full ingestion."""
        document = await self.neo4j_client.create_document(
            title=input_data.title,
            content_type=input_data.content_type,
            raw_content=input_data.raw_content,
            container_tag=input_data.container_tag,
            metadata=input_data.metadata,
            source_url=input_data.source_url,
            file_path=input_data.file_path,
        )
        result = await self.process_document(document.id)
        return {
            "document": result["document"],
            "chunk_count": result["chunk_count"],
            "memory_count": result["memory_count"],
        }

    async def process_document(
        self, document_id: str
    ) -> dict[str, Document | int]:
        """Runs pipeline stages for an existing document with parallel LLM calls."""
        document = await self.neo4j_client.get_document(document_id)
        if not document:
            raise ValueError(f"Document not found: {document_id}")

        try:
            await self._set_status(document.id, DocumentStatus.EXTRACTING)
            extractor = self._get_extractor(document.content_type)
            extracted_text = await extractor.extract(document)
            document = await self.neo4j_client.update_document(
                document.id,
                raw_content=extracted_text,
                status=DocumentStatus.EXTRACTING,
            )

            await self._set_status(document.id, DocumentStatus.CHUNKING)
            chunks = self.chunking_service.chunk(
                document=document,
                content=extracted_text,
            )

            await self._set_status(document.id, DocumentStatus.EXTRACTING_MEMORIES)
            filter_prompt = await self.neo4j_client.get_container_filter_prompt(
                document.container_tag
            )
            
            # Parallel extraction of memories from chunks with concurrency limit
            extracted_memories = await self._extract_memories_parallel(
                chunks, filter_prompt
            )

            await self._set_status(document.id, DocumentStatus.EMBEDDING)
            # Parallel embedding generation for chunks and memories
            chunk_contents = [chunk.content for chunk in chunks]
            memory_contents = [memory.content for memory in extracted_memories]
            
            chunk_embeddings, memory_embeddings = await asyncio.gather(
                self.embedding_service.generate_embeddings(chunk_contents),
                self.embedding_service.generate_embeddings(memory_contents),
            )

            await self._set_status(document.id, DocumentStatus.INDEXING)
            
            # Batch create chunks
            if chunks:
                await self.neo4j_client.create_chunks(
                    [
                        {
                            "content": chunk.content,
                            "chunk_index": chunk.chunk_index,
                            "container_tag": document.container_tag,
                            "source_doc_id": document.id,
                            "metadata": (
                                chunk.metadata.model_dump_json()
                                if hasattr(chunk.metadata, "model_dump_json")
                                else str(chunk.metadata)
                            ),
                            "embedding": chunk_embeddings[i],
                        }
                        for i, chunk in enumerate(chunks)
                    ]
                )

            # Prepare batch inputs for memory creation
            memory_inputs: List[MemoryBatchInput] = []
            for extracted_memory, embedding in zip(extracted_memories, memory_embeddings):
                if not extracted_memory or not embedding:
                    continue
                memory_inputs.append(
                    MemoryBatchInput(
                        content=extracted_memory.content,
                        memory_type=extracted_memory.memory_type,
                        container_tag=document.container_tag,
                        confidence=extracted_memory.confidence,
                        embedding=embedding,
                        source_doc_id=document.id,
                        valid_from=extracted_memory.valid_from,
                        valid_to=extracted_memory.valid_to,
                    )
                )

            # Batch create memories
            memory_ids: List[str] = []
            if memory_inputs:
                memory_ids = await batch_create_memories(
                    self.neo4j_client, memory_inputs
                )

            await self._set_status(document.id, DocumentStatus.COMPLETED)

            return {
                "document": document,
                "chunk_count": len(chunks),
                "memory_count": len(memory_ids),
            }

        except Exception as e:
            await self._set_status(document.id, DocumentStatus.FAILED)
            raise RuntimeError(f"Pipeline failed for document {document_id}: {e}") from e

    def _get_extractor(self, content_type: ContentType) -> Extractor:
        """Get appropriate extractor for content type."""
        match content_type:
            case ContentType.CONVERSATION:
                return ConversationExtractor()
            case ContentType.URL:
                return UrlExtractor()
            case ContentType.TEXT:
                return TextExtractor()
            case _:
                return TextExtractor()

    async def _set_status(self, document_id: str, status: DocumentStatus) -> None:
        """Update document status."""
        await self.neo4j_client.update_document(document_id, status=status)

    async def _extract_memories_parallel(
        self,
        chunks: List[ChunkPayload],
        filter_prompt: Optional[str],
        max_concurrency: int = 5,
    ) -> List[ExtractedMemory]:
        """Extract memories from chunks in parallel with concurrency limit."""
        async def extract_from_chunk(chunk: ChunkPayload) -> List[ExtractedMemory]:
            return await self.memory_extractor_service.extract_memories(
                chunk.content, filter_prompt
            )

        results = await gather_with_limit(
            chunks, extract_from_chunk, max_concurrency
        )
        
        # Flatten list of lists
        return [memory for sublist in results for memory in sublist]
